src/script_opt/Inline.cc

```diff
@@ -120,33 +122,171 @@ void Inliner::Analyze() {
+        // it takes more time to compile it than to just run it via the
+        // interpreter, it's a lose.
+        static std::string zeek_init_name = "zeek_init";
+        if ( func->Name() == zeek_init_name )
+            continue;
+
+        const auto& body = f.Body();
+
+        if ( func->GetKind() == Func::SCRIPT_FUNC && func->GetBodies().size() > 1 ) {
+            if ( event_handlers.count(func) == 0 )
+                event_handlers[func] = 1;
+            else
+                ++event_handlers[func];
+            ASSERT(body_to_info.count(body.get()) == 0);
+            body_to_info[body.get()] = i;
+        }
+    }
+
+    for ( auto& e : event_handlers ) {
+        auto func = e.first;
+        auto& bodies = func->GetBodies();
+        if ( bodies.size() != e.second )
+            // It's potentially unsound to inline some-but-not-all event
+            // handlers, because doing so may violate &priority's. We
+            // could do the work of identifying such instances and only
+            // skipping those, but given that ZAM is feature-complete
+            // the mismatch here should only arise when using restrictions
+            // like --optimize-file, which likely aren't the common case.
+            continue;
+
+        CoalesceEventHandlers({NewRef{}, func}, bodies, body_to_info);
+    }
+}
+
+void Inliner::CoalesceEventHandlers(ScriptFuncPtr func, const std::vector<Func::Body>& bodies,
+                                    const BodyInfo& body_to_info) {
+    auto merged_body = make_intrusive<StmtList>();
+    auto oi = merged_body->GetOptInfo();
+
+    auto& params = func->GetType()->Params();
+    auto nparams = params->NumFields();
+    size_t init_frame_size = static_cast<size_t>(nparams);
+
+    PreInline(oi, init_frame_size);
+
+    // We pattern the new (alternate) body off of the first body.
+    auto& b0 = func->GetBodies()[0].stmts;
+    auto b0_info = body_to_info.find(b0.get());
+    ASSERT(b0_info != body_to_info.end());
+    auto& info0 = funcs[b0_info->second];
+    auto& scope0 = info0.Scope();
+    auto& vars = scope0->OrderedVars();
+
+    // We need to create a new Scope. Otherwise, when inlining the first
+    // body the analysis of identifiers gets confused regarding whether
+    // a given identifier represents the outer instance or the inner.
+    auto empty_attrs = std::make_unique<std::vector<AttrPtr>>();
+    push_scope(scope0->GetID(), std::move(empty_attrs));
+
+    std::vector<IDPtr> param_ids;
+
+    for ( auto i = 0; i < nparams; ++i ) {
+        auto& vi = vars[i];
+        // We use a special scope name so that when debugging issues we can
+        // see that a given variable came from coalescing event handlers.
+        auto p = install_ID(vi->Name(), "<event>", false, false);
+        p->SetType(vi->GetType());
+        param_ids.push_back(std::move(p));
+    }
+
+    auto new_scope = pop_scope();
+    auto orig_scope = func->GetScope();
+    func->SetScope(new_scope);
```

Comment:
This modifies the original function? I presume that's what wanted (though it's a bit surprising). Merging as is, but maybe worth another look.

Reply:
Good catch - that's vestigial. Just pushed a fixup.

Reply:
I've picked the fixup last minute/second during the merge. I'll repush your branch and remove it from there so GitHub hopefully shows this PR as merged.